src/plugins/kibana/server/routes/api/ingest/register_bulk.js

```diff
@@ -0,0 +1,98 @@
+import { Promise } from 'bluebird';
+import { parse, transform } from 'csv';
+import _ from 'lodash';
+import hi from 'highland';
+import { patternToIngest } from '../../../../common/lib/convert_pattern_and_ingest_name';
+import { PassThrough } from 'stream';
+import bulkRequestSchema from '../../../lib/schemas/bulk_request_schema';
+
+export function registerBulk(server) {
+  server.route({
+    path: '/api/kibana/{id}/_bulk',
+    method: 'POST',
+    config: {
+      payload: {
+        output: 'stream',
+        maxBytes: 1024 * 1024 * 1024
+      },
+      validate: {
+        payload: bulkRequestSchema
+      }
+    },
+    handler: function (req, reply) {
+      const boundCallWithRequest = _.partial(server.plugins.elasticsearch.callWithRequest, req);
+      const indexPattern = req.params.id;
+      const usePipeline = req.payload.pipeline;
+      const csv = req.payload.csv;
+      const fileName = csv.hapi.filename;
+      const responseStream = new PassThrough();
+      const parser = parse({
+        columns: true,
+        auto_parse: true,
+        delimiter: _.get(req.payload, 'delimiter', ',')
+      });
+
+      let currentLine = 2; // Starts at 2 since we parse the header separately
+
+      responseStream.write('[');
+      csv.pipe(parser);
+
+      hi(parser)
```

Comment:
:+1: to not trying to build a third party api in this initial implementation, but that shouldn't mean we don't write tests to cover the code paths we create.  Again, I go back to my product vs code thing from before, the vast majority of our code isn't publicly consumable (or at least we don't want people to use it directly), but that doesn't mean we don't benefit from testing it.  Those ~24 code paths aren't accidental or anything, we're not talking about accounting for every possible input type and stuff, we're just talking about testing the code paths that are written.

Things like batching size, parallelization, and back pressure are handled by highland because the way you've coded this, but any developer at any time could make even trivial changes to this file that could break any or all of those things in subtle ways that could have effects we couldn't possibly anticipate through automated testing.  That's why we should assert that these things are being setup/configured the way they need to be rather than hoping our tests will catch the unexpected effects of breaking them.

That doesn't mean we need to re-test all of the capabilities of highland, though. If we are comfortable assuming that highland will handle those things for us so long as we set them or configure them properly (which I am, and it sounds like you are), then we would only need to assert that we're configuring them properly.  Those assertions also help codify important decisions that are made at implementation time, like the choice to do a lower batch size.

As for the rest of those things, I didn't mean to imply that you weren't testing for them, just that my 24 code paths number wasn't even considering those things.  Many of those things are definitely best verified through functional tests just as many others are best verified through unit tests.  You probably are verifying a bunch of them already.

To be honest, I think a boatload of context is being lost in this comment chain.  Want to zoom tomorrow?  Maybe we could bring in some other people to get some fresh perspectives as well?